org.eclipse.tm4e.ui/src/main/java/org/eclipse/tm4e/ui/text/TMPresentationReconciler.java

```diff
@@ -420,7 +422,21 @@ private void themeChange(ITokenProvider oldTheme, ITokenProvider newTheme, IDocu
 	@Override
 	public void install(ITextViewer viewer) {
 		Assert.isNotNull(viewer);
-
+		
+		if (this.grammar == null && viewer.getDocument() != null) {
+			try {
+				ContentTypeInfo info = ContentTypeHelper.findContentTypes(viewer.getDocument());
+				if (info != null) {
+					this.grammar = findGrammar(info);
+				}
+			} catch (CoreException e) {
+				TMUIPlugin.getDefault().getLog().log(new Status(IStatus.ERROR, TMUIPlugin.PLUGIN_ID, e.getMessage(), e));
+			}
+			if (this.grammar == null) {
+				throw new IllegalArgumentException("No TextMate Grammar for provided editor/viewer/document.");

```

Comment:
My fear is that with current Eclipse Oxygen, Generic Editor will not work if you have not configured the grammar, isn't it?

If it that perhaps we should not throw this error if editor is a Generic Editor with version < M4 (when your patch will be accepted).